extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessor.java

```diff
@@ -153,188 +161,118 @@ public List<WritableFrameChannel> outputChannels()
-          // if they have the same partition key
-          // keep adding them after checking
-          // guardrails
-          objectsOfASingleRac.add(currentRow);
-          if (objectsOfASingleRac.size() > maxRowsMaterialized) {
-            throw new MSQException(new TooManyRowsInAWindowFault(
-                objectsOfASingleRac.size(),
-                maxRowsMaterialized
-            ));
-          }
-        } else {
-          // key change noted
-          // create rac from the rows seen before
-          // run the operators on these rows and columns
-          // clean up the object to hold the new rows only
-          if (objectsOfASingleRac.size() > maxRowsMaterialized) {
-            throw new MSQException(new TooManyRowsInAWindowFault(
-                objectsOfASingleRac.size(),
-                maxRowsMaterialized
-            ));
-          }
-          RowsAndColumns rac = MapOfColumnsRowsAndColumns.fromResultRow(
-              objectsOfASingleRac,
-              frameReader.signature()
-          );
-          runAllOpsOnSingleRac(rac);
-          objectsOfASingleRac.clear();
-          outputRow = currentRow.copy();
-          return ReturnOrAwait.runAgain();
-        }
-        frameCursor.advance();
+    }
+
+    // Scenario 2: All window functions in the query have OVER() clause with a PARTITION BY
+    if (frameCursor == null || frameCursor.isDone()) {
+      if (readableInputs.isEmpty()) {
+        return ReturnOrAwait.awaitAll(1);
+      } else if (inputChannel.canRead()) {
+        final Frame frame = inputChannel.read();
+        frameCursor = FrameProcessors.makeCursor(frame, frameReader);
+        makeRowSupplierFromFrameCursor();
+      } else if (inputChannel.isFinished()) {
+        // Handle any remaining data.
+        batchOfRacsPendingProcessing.addAll(currentBatchOfRacs);
+        processBatchOfRacsPendingProcessing();
+        return ReturnOrAwait.returnObject(Unit.instance());
+      } else {
+        return ReturnOrAwait.runAgain();
       }
     }
-    return ReturnOrAwait.runAgain();
-  }
 
-  /**
-   * @param singleRac Use this {@link RowsAndColumns} as a single input for the operators to be run
-   */
-  private void runAllOpsOnSingleRac(RowsAndColumns singleRac)
-  {
-    Operator op = new Operator()
-    {
-      @Nullable
-      @Override
-      public Closeable goOrContinue(Closeable continuationObject, Receiver receiver)
-      {
-        receiver.push(singleRac);
-        if (singleRac.numRows() > maxRowsMaterialized) {
-          throw new MSQException(new TooManyRowsInAWindowFault(singleRac.numRows(), maxRowsMaterialized));
+    while (!frameCursor.isDone()) {
+      final ResultRow currentRow = rowSupplierFromFrameCursor.get();
+      if (outputRow == null) {
+        outputRow = currentRow;
+        currentBatchOfRacs.add(currentRow);
+      } else if (comparePartitionKeys(outputRow, currentRow, partitionColumnNames)) {
```

Comment:
@LakshSingla Step 3 is incorrectly demonstrated in the above table.

Rows are moved from `currentBatchOfRacs` to `batchOfRacsPendingProcessing` only in 2 scenarios:
1. Input channel is finished.
2. We encounted a PARTITION BY key change.

Step 3 doesn't fall under the above 2 scenarios.

Above example would look like following:
| Stream Element  |  batchOfRacsPendingProcessing | currentBatchOfRacs  |  Notes |
|---|---|---|---|
|  a | -  | a  | -  |
|  b | a  | b  | -  |
|  b | -  | b, b  | Flush a.  |
|  b | -  | b, b, b  | Error out because `currentBatchOfRacs.size() + batchOfRacsPendingProcessing.size() > maxRowsMaterialized`  |


```java
     if (outputRow == null) {
        // code
      } else if (comparePartitionKeys(outputRow, currentRow, partitionColumnNames)) {
        // code
          processBatchOfRacsPendingProcessing();
        }
      } else {
        processBatchOfRacsPendingProcessingIfNeeded();
        // code
      }
```
Notice `processBatchOfRacsPendingProcessing()` vs `processBatchOfRacsPendingProcessingIfNeeded()` in the above flow. (in case the confusion stemmed because of this)

Hope this clarifies.

Reply:
Thanks!